### ISSUE TITLE
Announcement: Driving towards zero 2.1 bugs by March 31 (ZBB)

### ISSUE THREAD

**Issue by karelz** [dotnet/corefx#28114]
As we are getting closer to .NET Core 2.1 [ship dates](https://github.com/dotnet/core/blob/master/roadmap.md#upcoming-ship-dates), we would like to **drive [2.1.0 bugs](https://github.com/dotnet/corefx/issues?utf8=%E2%9C%93&q=is%3Aissue%20is%3Aopen%20milestone%3A2.1.0) close to 0 by March 31** = ZBB (zero bug bounce) date.
Note: We will keep fixing 2.1 issues after that, until the final release (exact dates are not announced yet), but hopefully not in large masses.

We would like to encourage and ask our contributors to help us drive towards that goal. **Be part of .NET Core 2.1 release!**

How can you, the CoreFX contributors, help?
1. If you're picking up issues to work on, **prefer issues marked as [2.1.0 bugs](https://github.com/dotnet/corefx/issues?utf8=%E2%9C%93&q=is%3Aissue%20is%3Aopen%20milestone%3A2.1.0)** (some are even [up-for-grabs](https://github.com/dotnet/corefx/issues?utf8=%E2%9C%93&q=is%3Aissue%20is%3Aopen%20milestone%3A2.1.0%20label%3Aup-for-grabs)).
2. If you spot something in [Future](https://github.com/dotnet/corefx/issues?utf8=%E2%9C%93&q=is%3Aissue%20is%3Aopen%20milestone%3AFuture) milestone that should be part of 2.1, please say so and explain why (loop in @karelz). Conversely if there's something in 2.1 that is not as significant as it seems.
    * We strive to deliver high-quality product, so **impactful bugs, compatibility, reliability and performance issues** are top priority. We want to be highly compatible with .NET Core 2.0.
    * Adding additional new APIs, fixing pre-existing bugs or fixing corner-case issues with lower impact and with existing workaround is lower priority, not tracked as 2.1. We will delay such contributions until master branch reopens for post-2.1 work (early April).
3. Issues with [no milestone](https://github.com/dotnet/corefx/issues?utf8=%E2%9C%93&q=is%3Aissue%20is%3Aopen%20no%3Amilestone) - we expect to set their milestone by end of week (part of ongoing "triage"), so if you see something must-have for 2.1 there, please say so and explain why (loop in @karelz).
4. Watch for upcoming official preview announcements (on [.NET blog](https://blogs.msdn.microsoft.com/dotnet/) or on [dotnet/Announcements repo](https://github.com/dotnet/announcements)) to help validate and dogfood the end-to-end .NET Core 2.1 experience before final release.
    * You can start with [.NET Core 2.1 Preview 1](https://blogs.msdn.microsoft.com/dotnet/2018/02/27/announcing-net-core-2-1-preview-1/) released on Feb 27.
    * You can also dogfood our daily builds - see [dogfooding steps](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/dogfooding.md)

Not surprisingly, to deliver overall end-to-end .NET Core 2.1 experience, similar ZBB efforts happen in our sister repos like [CoreCLR](https://github.com/dotnet/coreclr), [core-setup](https://github.com/dotnet/core-setup), etc. - we will welcome help in those as well, if you're interested.

Twitter announcement: https://twitter.com/ziki_cz/status/974383981436088320

**Comment by karelz** [comment]
See our progress on bugs in 2018 (as of yesterday 3/14 13:38 PT): https://twitter.com/ziki_cz/status/974384999423082496

**Comment by tarekgh** [comment]
@karelz why we don't use https://github.com/dotnet/announcements/issues for the announcements instead? it looks this would be a better place for the whole release.

**Comment by tarekgh** [comment]
CC @AlexGhiondea

**Comment by karelz** [comment]
cc'ing recent contributors to CoreFX repo (in random order):
@AlexRadch @kasper3 @AppBeat @jswolf19 @Jlalond @cwe1ss @sherlock1982 @mattpwhite @joelbraun  @sdmaclea @fiigii @tmds @MarcoRossignoli @JonHanna @sylveon @petarpetrovt @RemiBou @omajid @kekekeks @hughbe @benjamin-hodgson @AlexG2017 @artkpv @clarkis117 @github-john-doe @nietras @baulig @JeffCyr @pepone @alpencolt @jbednar79 @Lroca88 @k0st1x

**Comment by karelz** [comment]
and more:
@OzieGamma @bording @svick @WinCPP @weitzhandler @Thiez @gfoidl @Douglas-Cleghorn @testfirstcoder @justinvp @KeithHenry @pentp @dajeffri @logiclrd @ah- @davidmatson @strajkovmsft @ayende @kassemsandarusi @philippgille @alfredmyers @diamondo25 @mqudsi @the-dwyer @TylerBrinkley @jcdickinson @madelson @LotharOfTheHillPeople @EgorBo @peterhpchen @TheSatoshiChiba @satano @fadamsen @JohnTortugo @zaytsev-victor @FSou1 @Drawaes @khellang @benaadams @lewurm @rogaps @sonnemaf

**Comment by karelz** [comment]
and more:
@FransBouma @Cronan @kgarlinski @lzybkr @cod7alex @adriangodong @stukselbax @reaction1989 @KostaVlev @qmfrederik @4creators @yvanin @deinok @jaredcnance @ektrah @brunolins16 @sh3ev @MrDesjardins @MaxHorstmann @jpelcis @wateret @jiria @hqueue @SimonCropp @supporterPL @teeram @poizan42

**Comment by AlexRadch** [comment]
I starting add tests for ReadOnlySequence longer than int.Max. It is second part of issue https://github.com/dotnet/corefx/issues/26768.

Also I created PR with new tests for ReadOnlySequence that it fail. https://github.com/dotnet/corefx/pull/28095 . If new tests are correct I will try to remove bugs in ReadOnlySequence to pass new tests.

**Comment by karelz** [comment]
Great, thanks! I will lock the discussion here to keep it scoped to updates to the announcement.
For specific suggestions, questions and concerns, please use existing issues or file new ones (feel free to tag me).

**Comment by karelz** [comment]
Progress on bugs: https://twitter.com/ziki_cz/status/976549475513544705
Highlights:
* Great progress on issues without milestone over last week - it will allow us to focus even more on 2.1 bugs now.
    * We went down to "just" 26 issues without milestone from 252 (+incoming)
    * We have only 6 issues without milestone are older than 1 week
* 1.5 weeks until 3/31 and plenty of 2.1 bugs to address - we plan to push on higher triage bar (nice-to-haves will be pushed to Future) and of course on more fixes getting in.

**Comment by karelz** [comment]
Progress on bugs: https://twitter.com/ziki_cz/status/977239890642391040
Highlights:
* Awesome progress on 2.1 bugs in both bug fixes and triage - we asked the teams to properly apply 2.1 bar and push nice-to-have, non-critical issues into Future.
    * Thanks all contributors for helping us with 2.1 bugs!!!
* Issues without milestone are under control, kept low. We have only 2 issues without milestone older than 1 week.
* We're still looking for more feedback - if you can try our daily bits, please do so!
    * Special ask for HttpClient coverage: It is backed by brand new C#-based implementation. It is more consistent across OSs and faster :). Any feedback will be greatly appreciated!

**Comment by karelz** [comment]
Progress on bugs: https://twitter.com/ziki_cz/status/979410567659118592 (last 3 days left)
Highlights:
* Great progress on 2.1 bugs and triage
    * We introduced Post-ZBB label for exceptions (currently just few bugs: [3](https://github.com/dotnet/corefx/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+milestone%3A2.1.0+label%3APost-ZBB) + [0](https://github.com/dotnet/coreclr/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+milestone%3A2.1.0+label%3APost-ZBB) + [0](https://github.com/dotnet/core-setup/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+milestone%3A2.1.0+label%3APost-ZBB) + [0](https://github.com/dotnet/wcf/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+milestone%3A2.1.0+label%3APost-ZBB)).
    * We will exclude bugs filed after 3/28 from ZBB
    * Per-repo ZBB query: **111** = [60](https://github.com/dotnet/corefx/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+milestone%3A2.1.0+created%3A%3C2018-03-29+-label%3APost-ZBB+) + [27](https://github.com/dotnet/coreclr/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+milestone%3A2.1.0+created%3A%3C2018-03-29+-label%3APost-ZBB+) + [20](https://github.com/dotnet/core-setup/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+milestone%3A2.1.0+created%3A%3C2018-03-29+-label%3APost-ZBB+) + [4](https://github.com/dotnet/wcf/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+milestone%3A2.1.0+created%3A%3C2018-03-29+-label%3APost-ZBB+)
* Issues without milestone are under control
    * We will exclude bugs filed after 3/28 from ZBB
    * Per-repo ZBB no-milestone query: **16** = [7](https://github.com/dotnet/corefx/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+no%3Amilestone+created%3A%3C2018-03-29) + [0](https://github.com/dotnet/coreclr/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+no%3Amilestone+created%3A%3C2018-03-29) + [1](https://github.com/dotnet/core-setup/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+no%3Amilestone+created%3A%3C2018-03-29) + [8](https://github.com/dotnet/wcf/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+no%3Amilestone+created%3A%3C2018-03-29)
* Overall 2.1 bugs (not just subject of ZBB): **118** = [67](https://github.com/dotnet/corefx/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+milestone%3A2.1.0) + [27](https://github.com/dotnet/coreclr/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+milestone%3A2.1.0) + [20](https://github.com/dotnet/core-setup/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+milestone%3A2.1.0) + [4](https://github.com/dotnet/wcf/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+milestone%3A2.1.0)
* Keep feedback coming - try our **[daily builds](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/dogfooding.md)** if you can (all work in master branch still flows into 2.1).

**Comment by karelz** [comment]
We've got very close to zero 2.1 ZBB bugs today - we're down to **just 21 issues for ZBB!** (excluding exceptions and bugs filed after 3/28):

* **2.1 ZBB bugs:** **21** = [18](https://github.com/dotnet/corefx/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+milestone%3A2.1.0+created%3A%3C2018-03-29+-label%3APost-ZBB) + [0](https://github.com/dotnet/coreclr/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+milestone%3A2.1.0+created%3A%3C2018-03-29+-label%3APost-ZBB) + [3](https://github.com/dotnet/core-setup/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+milestone%3A2.1.0+created%3A%3C2018-03-29+-label%3APost-ZBB) + [0](https://github.com/dotnet/wcf/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+milestone%3A2.1.0+created%3A%3C2018-03-29+-label%3APost-ZBB)
    * We have quite a few of these in PRs, so we expect to be even closer to ZERO on Monday
* **ZERO** issues without milestone created after 3/28: **0** = [0](https://github.com/dotnet/corefx/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+no%3Amilestone+created%3A%3C2018-03-29) + [0](https://github.com/dotnet/coreclr/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+no%3Amilestone+created%3A%3C2018-03-29) + [0](https://github.com/dotnet/core-setup/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+no%3Amilestone+created%3A%3C2018-03-29) + [0](https://github.com/dotnet/wcf/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+no%3Amilestone+created%3A%3C2018-03-29)
* Exceptions for ZBB are fairly under control: **36** = [12](https://github.com/dotnet/corefx/issues?q=is%3Aissue+is%3Aopen+milestone%3A2.1.0+label%3APost-ZBB) + [14](https://github.com/dotnet/coreclr/issues?q=is%3Aissue+is%3Aopen+milestone%3A2.1.0+label%3APost-ZBB) + [10](https://github.com/dotnet/core-setup/issues?q=is%3Aissue+is%3Aopen+milestone%3A2.1.0+label%3APost-ZBB) + [0](https://github.com/dotnet/wcf/issues?q=is%3Aissue+is%3Aopen+milestone%3A2.1.0+label%3APost-ZBB)
    * All remaining 2.1 bugs: **72** = [37](https://github.com/dotnet/corefx/issues?q=is%3Aopen+is%3Aissue+milestone%3A2.1.0) + [19](https://github.com/dotnet/coreclr/issues?q=is%3Aopen+is%3Aissue+milestone%3A2.1.0) + [16](https://github.com/dotnet/core-setup/issues?q=is%3Aopen+is%3Aissue+milestone%3A2.1.0) + [0](https://github.com/dotnet/wcf/issues?q=is%3Aopen+is%3Aissue+milestone%3A2.1.0)

Branch status:
* We did one integration master -> release/2.1 early today (Fri 3/30). We plan another one on Mon 4/2.
    * 2.1 bug fixes will be cherry picked from master and manually ported, if they meet the bar (i.e. impactful for 2.1 - we will start shiproom process for that shortly).
* We will keep master branch locked down for 2.1 work for another ~2 weeks (just in case + we need to prep infrastructure for 2.2 work).
    * Risky and disruptive changes will be delayed until master branch is opened for 2.2 work (in ~2 weeks).

**Comment by karelz** [comment]
We've hit ZERO 2.1 ZBB bugs! (excluding exceptions and bugs filed after 3/28):

* **2.1 ZBB bugs:** **0** = [0](https://github.com/dotnet/corefx/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+milestone%3A2.1.0+created%3A%3C2018-03-29+-label%3APost-ZBB) + [0](https://github.com/dotnet/coreclr/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+milestone%3A2.1.0+created%3A%3C2018-03-29+-label%3APost-ZBB) + [0](https://github.com/dotnet/core-setup/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+milestone%3A2.1.0+created%3A%3C2018-03-29+-label%3APost-ZBB) + [0](https://github.com/dotnet/wcf/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+milestone%3A2.1.0+created%3A%3C2018-03-29+-label%3APost-ZBB)
    * Also no issues without milestone created before 3/29: **0** = [0](https://github.com/dotnet/corefx/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+no%3Amilestone+created%3A%3C2018-03-29) + [0](https://github.com/dotnet/coreclr/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+no%3Amilestone+created%3A%3C2018-03-29) + [0](https://github.com/dotnet/core-setup/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+no%3Amilestone+created%3A%3C2018-03-29) + [0](https://github.com/dotnet/wcf/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+no%3Amilestone+created%3A%3C2018-03-29)
* Exceptions for ZBB are fairly under control: **43** = [19](https://github.com/dotnet/corefx/issues?q=is%3Aissue+is%3Aopen+milestone%3A2.1.0+label%3APost-ZBB) + [12](https://github.com/dotnet/coreclr/issues?q=is%3Aissue+is%3Aopen+milestone%3A2.1.0+label%3APost-ZBB) + [12](https://github.com/dotnet/core-setup/issues?q=is%3Aissue+is%3Aopen+milestone%3A2.1.0+label%3APost-ZBB) + [0](https://github.com/dotnet/wcf/issues?q=is%3Aissue+is%3Aopen+milestone%3A2.1.0+label%3APost-ZBB)
    * All remaining 2.1 bugs: **56** = [26](https://github.com/dotnet/corefx/issues?q=is%3Aopen+is%3Aissue+milestone%3A2.1.0) + [15](https://github.com/dotnet/coreclr/issues?q=is%3Aopen+is%3Aissue+milestone%3A2.1.0) + [15](https://github.com/dotnet/core-setup/issues?q=is%3Aopen+is%3Aissue+milestone%3A2.1.0) + [0](https://github.com/dotnet/wcf/issues?q=is%3Aopen+is%3Aissue+milestone%3A2.1.0)

**Comment by karelz** [comment]
Quick update:
* We decided to keep master branch in lock down for 2.1 release until Fri 4/13. We will merge only PRs which are critical for 2.1 release.
    * Controversial or risky changes without significant value will be on hold until the branch is reopened for 2.2 work (~Mon 4/16).
* We plan (another) last integration master->release/2.1 on Fri 4/13.

We created 2.2.0 (placeholder) milestone in our repos core-* & wcf to track work in next major release. We will move important issues from Future milestone to 2.2.0 milestone as priorities and dates for the next release become more clear.
